scripts/cleanup_analysis_entries.py

```diff
@@ -0,0 +1,259 @@
+
+    print(f'{len(path_existence["valid"])} valid paths, {len(path_existence["invalid"])} invalid paths')
+
+    print('Analysis paths with no file discovered:', path_existence['invalid'])
+
+
+    # now do more stuff to check:
+    #   analysis-entries we have to delete because files don't exist
+    #   samples that DID have analysis and now don't
+    if dry_run:
+        print(f'Ready for ids to delete: {ids_to_delete}')
+    else:
+        # delete some analysis entries
+        pass
+
+def get_analyses_for_datasets(datasets, analysis_type) -> list[dict]:
+    # get all analysis-entries 
+    return aapi.query_analyses(
+            AnalysisQueryModel(
+                projects=datasets,
+                type=AnalysisType(analysis_type),
+                # status=AnalysisStatus('completed'),
+            )
+        )
+
+def find_invalid_analyses(analyses: list[dict]):
+    """
+    
+    
+    """
+    # flag analyses in a bad state
+    pass
+
+
+def get_path_components_from_path(path):
+    """
+    Return the {bucket_name}, {dataset}, and {subdir} for GS only paths
+
+    >>> get_bucket_name_path_from_path('gs://cpg-dataset-main/subfolder/subfolder2/my.cram')
+    {bucket_name:'cpg-dataset-main', dataset:'dataset', subdir:'subfolder/subfolder2'}
+    """
+    if path.startswith('gs://'):
+        short_path = path.removeprefix('gs://').split('/', maxsplit=1)
+        bucket_name = short_path[0]
+
+        dataset_with_bucket_type = bucket_name.removeprefix('cpg-')
+        bucket_type = dataset_with_bucket_type.split('-')[-1]
+        dataset = dataset_with_bucket_type.removesuffix(bucket_type)[:-1]
+
+        file_name = short_path[1].split('/')[-1]
+        subdir = short_path[1].removesuffix(file_name)[:-1]
+
+    else:
+        raise warnings.warn(f'Analysis path {path} not bucket storage.')
+    
+    return {'bucket_name':bucket_name, 'dataset':dataset, 'subdir':subdir}
+
+def deduplicate_analyses(analyses: list[dict]) -> tuple[list[dict], list[int], list[list]]:
+    # Check which analysis IDs point to the same analysis output path
+    # Keep the smallest ID pointing to a given analysis output, duplicates can be scheduled for deletion
+
+    all_analyses = defaultdict(str)
+    # map each analysis id to its output path
+    for analysis in analyses:
+        all_analyses[analysis['id']]=analysis['output']
+    
+    # map each output path to the analysis entries which point to it
+    reversed_analyses_dict = {}
+    for k, v in all_analyses.items():
+        reversed_analyses_dict.setdefault(v, set()).add(k)
+
+    duplicate_analyses = [vals for vals in reversed_analyses_dict.values() if len(vals)>1]
+    
+    # keep the first entry pointing to a path, duplicates can be deleted
+    duplicate_analyses_ids = []
+    for duplicate_ids in duplicate_analyses:
+        duplicate_analyses_ids.append(sorted(duplicate_ids)[1:])
+
+    ids_to_delete = [e for sublist in duplicate_analyses_ids for e in sublist]
```

Comment:
```suggestion
    ids_to_delete = {}
    for duplicate_ids in duplicate_analyses:
        ids_to_delete.update(sorted(duplicate_ids)[1:])
```

Alternatively the `itertools.chain.from_iterable` method is a neat way to avoid this double layered list comp

```
from itertools import chain
ids_to_delete = chain.from_iterable(duplicate_analyses_ids)
```